File: src/org/simbrain/network/synapse_update_rules/PfisterGerstner2006Rule.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.network.synapse_update_rules;

import org.simbrain.network.core.Synapse;
import org.simbrain.network.core.SynapseUpdateRule;
import org.simbrain.network.gui.UserParameter;


/**
 * Implementation of the model described by Pfister, J-P, Gerstner, W: 
 * Triplets of Spikes in a Model of Spike Timing-Dependent Plasticity. 
 * J. Neurosci. 26, 9673–9682 (2006).
 * 
 * Only works if source and target neurons are spiking neurons.
 * 
 * @author Oliver J. Coleman
 */
public class PfisterGerstner2006Rule extends SynapseUpdateRule implements Cloneable {
	@UserParameter(label="Tau+", description="Decay rate for r1 trace", minimumValue=1, maximumValue=100, defaultValue="16.8", order=0)
    protected double tauPlus = 16.8;

	@UserParameter(label="Tau x", description="Decay rate for r2 trace", minimumValue=1, maximumValue=100, defaultValue="1.0", order=1)
    protected double tauX = 1.0;

	@UserParameter(label="Tau-", description="Decay rate for o1 trace", minimumValue=1, maximumValue=100, defaultValue="33.7", order=2)
    protected double tauNeg = 33.7;

	@UserParameter(label="Tau y", description="Decay rate for o2 trace", minimumValue=1, maximumValue=100, defaultValue="48.0", order=3)
    protected double tauY = 48.0;

	@UserParameter(label="A2+", description="Amplitude of the weight change for a pre-post spike pair.", minimumValue=0, maximumValue=0.1, defaultValue="0.003", order=4)
    protected double a2P = 0.0046;

	@UserParameter(label="A2-", description="Amplitude of the weight change for a post-pre spike pair.", minimumValue=0, maximumValue=0.1, defaultValue="0.003", order=5)
    protected double a2N = 0.003;

	@UserParameter(label="A3+", description="Amplitude of the triplet term for potentiation.", minimumValue=0, maximumValue=0.1, defaultValue="0.0091", order=6)
    protected double a3P = 0.0091;
    
	@UserParameter(label="A3-", description="Amplitude of the triplet term for depression.", minimumValue=0, maximumValue=0.1, defaultValue="0.0", order=7)
    protected double a3N = 0;
	
	
    // Spike traces.
	private double r1, r2, o1, o2;
	// Cached multipliers for trace decays.
	private double tauPlusMult, tauXMult, tauNegMult, tauYMult;
    
	
    @Override
    public void init(Synapse synapse) {
    	tauPlusMult = 1 / tauPlus;
        tauXMult = 1 / tauX;
        tauNegMult = 1 / tauNeg;
        tauYMult = 1 / tauY;
    }

    
    @Override
    public String getName() {
        return "Pfister and Gerstner, 2006";
    }

    
    @Override
    public SynapseUpdateRule deepCopy() {
    	// We're only using primitive fields so clone() works.
    	try {
			return (PfisterGerstner2006Rule) this.clone();
		} catch (CloneNotSupportedException e) {
			throw new RuntimeException(e);
		}
    }
    
    
    @Override
    public void update(Synapse synapse) {
    	// Time step in ms.
    	final double timeStep = synapse.getNetwork().getTimeStep();
    	final boolean preSpiked = synapse.getSource().isSpike();
    	final boolean postSpiked = synapse.getTarget().isSpike();
    		
		// Need current values for these traces for strength update equations below.
		final double r2p = r2;
		final double o2p = o2;

		// Update trace values.
		if (preSpiked) {
			r1 = 1;
			r2 = 1;
		}
		else {
			r1 -= r1 * tauPlusMult * timeStep;
			r2 -= r2 * tauXMult * timeStep;
		}
		if (postSpiked) {
			o1 = 1;
			o2 = 1;
		}
		else {
			o1 -= o1 * tauNegMult * timeStep;
			o2 -= o2 * tauYMult * timeStep;
		}
		
		// Update efficacy if a pre or post spike occurred.
		if (preSpiked) {
			synapse.setStrength(synapse.getStrength() - o1 * (a2N + a3N * r2p));
		}
		
		if (postSpiked) {
			synapse.setStrength(synapse.getStrength() + r1 * (a2P + a3P * o2p));
		}
    }


	/**
	 * @return Decay rate for r1 trace.
	 */
	public double getTauPlusDecay() {
		return 1 / tauPlusMult;
	}


	/**
	 * @param tauPlus Decay rate for r1 trace.
	 */
	public void setTauPlusDecay(double tauPlus) {
		this.tauPlusMult = 1 / tauPlus;
	}


	/**
	 * @return Decay rate for r2 trace.
	 */
	public double getTauXDecay() {
		return 1 / tauXMult;
	}


	/**
	 * @param tauX Decay rate for r2 trace.
	 */
	public void setTauXDecay(double tauX) {
		this.tauXMult = 1 / tauX;
	}


	/**
	 * @return Decay rate for o1 trace.
	 */
	public double getTauNegDecay() {
		return 1 / tauNegMult;
	}


	/**
	 * @param tauNeg Decay rate for o1 trace.
	 */
	public void setTauNegDecay(double tauNeg) {
		this.tauNegMult = 1 / tauNeg;
	}


	/**
	 * @return Decay rate for o2 trace.
	 */
	public double getTauYDecay() {
		return 1 / tauYMult;
	}


	/**
	 * @param tauY Decay rate for o2 trace.
	 */
	public void setTauYDecay(double tauY) {
		this.tauYMult = 1 / tauY;
	}


	/**
	 * @return Amplitude of the weight change for a pre-post spike pair.
	 */
	public double getA2P() {
		return a2P;
	}


	/**
	 * @param a2p Amplitude of the weight change for a pre-post spike pair.
	 */
	public void setA2P(double a2p) {
		a2P = a2p;
	}


	/**
	 * @return Amplitude of the weight change for a post-pre spike pair.
	 */
	public double getA2N() {
		return a2N;
	}


	/**
	 * @param Amplitude of the weight change for a post-pre spike pair.
	 */
	public void setA2N(double a2n) {
		a2N = a2n;
	}


	/**
	 * @return Amplitude of the triplet term for potentiation.
	 */
	public double getA3P() {
		return a3P;
	}


	/**
	 * @param Amplitude of the triplet term for potentiation.
	 */
	public void setA3P(double a3p) {
		a3P = a3p;
	}


    /**
	 * @return Amplitude of the triplet term for depression.
	 */
	public double getA3N() {
		return a3N;
	}


	/**
	 * @param a3n Amplitude of the triplet term for depression.
	 */
	public void setA3N(double a3n) {
		a3N = a3n;
	}
}

File: src/main/kotlin/org/simbrain/network/learningrules/PfisterGerstner2006Rule.kt
package org.simbrain.network.learningrules

import org.simbrain.network.core.*
import org.simbrain.network.util.EmptyMatrixData
import org.simbrain.network.util.EmptyScalarData
import org.simbrain.network.util.SpikingMatrixData
import org.simbrain.util.UserParameter

/**
 * Implementation of the model described by Pfister, J-P, Gerstner, W: Triplets
 * of Spikes in a Model of Spike Timing-Dependent Plasticity. J. Neurosci. 26,
 * 9673–9682 (2006).
 *
 *
 * Only works if source and target neurons are spiking neurons.
 *
 * @author Oliver J. Coleman
 */
class PfisterGerstner2006Rule : SynapseUpdateRule<EmptyScalarData, EmptyMatrixData>(), Cloneable {
    @UserParameter(
        label = "Tau+",
        description = "Decay rate for r1 trace",
        minimumValue = 1.0,
        maximumValue = 100.0,
        increment = .1,
        order = 0
    )
    protected var tauPlus: Double = 16.8

    @UserParameter(
        label = "Tau x",
        description = "Decay rate for r2 trace",
        minimumValue = 1.0,
        maximumValue = 100.0,
        increment = .1,
        order = 1
    )
    protected var tauX: Double = 1.0

    @UserParameter(
        label = "Tau-",
        description = "Decay rate for o1 trace",
        minimumValue = 1.0,
        maximumValue = 100.0,
        increment = .1,
        order = 2
    )
    protected var tauNeg: Double = 33.7

    @UserParameter(
        label = "Tau y",
        description = "Decay rate for o2 trace",
        minimumValue = 1.0,
        maximumValue = 100.0,
        increment = .1,
        order = 3
    )
    protected var tauY: Double = 48.0

    /**
     * @return Amplitude of the weight change for a pre-post spike pair.
     */
    /**
     * @param a2p Amplitude of the weight change for a pre-post spike pair.
     */
    @UserParameter(
        label = "A2+",
        description = "Amplitude of the weight change for a pre-post spike pair.",
        minimumValue = 0.0,
        maximumValue = 0.1,
        increment = .1,
        order = 4
    )
    var a2P: Double = 0.0046

    /**
     * @return Amplitude of the weight change for a post-pre spike pair.
     */
    /**
     * @param a2n Amplitude of the weight change for a post-pre spike pair.
     */
    @UserParameter(
        label = "A2-",
        description = "Amplitude of the weight change for a post-pre spike pair.",
        minimumValue = 0.0,
        maximumValue = 0.1,
        increment = .1,
        order = 5
    )
    var a2N: Double = 0.003

    /**
     * @return Amplitude of the triplet term for potentiation.
     */
    /**
     * @param a3p Amplitude of the triplet term for potentiation.
     */
    @UserParameter(
        label = "A3+",
        description = "Amplitude of the triplet term for potentiation.",
        minimumValue = 0.0,
        maximumValue = 0.1,
        increment = .1,
        order = 6
    )
    var a3P: Double = 0.0091

    /**
     * @return Amplitude of the triplet term for depression.
     */
    /**
     * @param a3n Amplitude of the triplet term for depression.
     */
    @UserParameter(
        label = "A3-",
        description = "Amplitude of the triplet term for depression.",
        minimumValue = 0.0,
        maximumValue = 0.1,
        increment = .1,
        order = 7
    )
    var a3N: Double = 0.0

    // Spike traces.
    private var r1 = 0.0
    private var r2 = 0.0
    private var o1 = 0.0
    private var o2 = 0.0

    // Cached multipliers for trace decays.
    private var tauPlusMult = 0.0
    private var tauXMult = 0.0
    private var tauNegMult = 0.0
    private var tauYMult = 0.0

    override fun init(synapse: Synapse) {
        tauPlusMult = 1 / tauPlus
        tauXMult = 1 / tauX
        tauNegMult = 1 / tauNeg
        tauYMult = 1 / tauY
    }

    override val name: String
        get() = "Pfister and Gerstner, 2006"

    override fun copy(): SynapseUpdateRule<*, *> {
        // We're only using primitive fields so clone() works.
        try {
            return clone() as PfisterGerstner2006Rule
        } catch (e: CloneNotSupportedException) {
            throw RuntimeException(e)
        }
    }

    context(Network)
    override fun apply(synapse: Synapse, data: EmptyScalarData) {
        val preSpiked = synapse.source.isSpike
        val postSpiked = synapse.target.isSpike

        // Need current values for these traces for strength update equations
        // below.
        val r2p = r2
        val o2p = o2

        // Update trace values.
        if (preSpiked) {
            r1 = 1.0
            r2 = 1.0
        } else {
            r1 -= r1 * tauPlusMult * timeStep
            r2 -= r2 * tauXMult * timeStep
        }
        if (postSpiked) {
            o1 = 1.0
            o2 = 1.0
        } else {
            o1 -= o1 * tauNegMult * timeStep
            o2 -= o2 * tauYMult * timeStep
        }

        // Update efficacy if a pre or post spike occurred.
        if (preSpiked) {
            synapse.strength = synapse.strength - o1 * (a2N + a3N * r2p)
        }

        if (postSpiked) {
            synapse.strength = synapse.strength + r1 * (a2P + a3P * o2p)
        }
    }

    // Note: This function has not been tested.
    // Note: Matrix implementation requires managing trace matrices for each connection
    // This is computationally intensive but doable for moderate sized networks
    context(Network)
    override fun apply(connector: Connector, dataHolder: EmptyMatrixData) {
        val weightMatrix = connector as? WeightMatrix ?: return
        val sourceNeuronArray = weightMatrix.source as? NeuronArray ?: return
        val targetNeuronArray = weightMatrix.target as? NeuronArray ?: return
        
        // Ensure both neuron arrays have spiking data
        val sourceSpikingData = sourceNeuronArray.dataHolder as? SpikingMatrixData ?: return
        val targetSpikingData = targetNeuronArray.dataHolder as? SpikingMatrixData ?: return
        
        // For matrix implementation, we would need to maintain trace matrices
        // This is a simplified version that applies the rule when spikes occur
        // A full implementation would require additional MatrixDataHolder for traces
        
        val sourceSpikes = sourceSpikingData.spikes
        val targetSpikes = targetSpikingData.spikes
        
        // Apply rule to each connection where spikes occur
        for (i in 0 until weightMatrix.weights.nrow()) { // target neurons
            for (j in 0 until weightMatrix.weights.ncol()) { // source neurons
                val preSpiked = sourceSpikes[j]
                val postSpiked = targetSpikes[i]
                
                if (preSpiked || postSpiked) {
                    // This is a simplified implementation - full version would need trace management
                    // For now, applying basic potentiation/depression based on spike timing
                    val currentWeight = weightMatrix.weights[i, j]
                    
                    if (preSpiked && postSpiked) {
                        // Both spiked - apply both LTP and LTD components
                        weightMatrix.weights[i, j] = currentWeight + (a2P - a2N)
                    } else if (preSpiked) {
                        // Only pre-spiked - apply LTD
                        weightMatrix.weights[i, j] = currentWeight - a2N
                    } else if (postSpiked) {
                        // Only post-spiked - apply LTP
                        weightMatrix.weights[i, j] = currentWeight + a2P
                    }
                }
            }
        }
    }

    var tauPlusDecay: Double
        /**
         * @return Decay rate for r1 trace.
         */
        get() = 1 / tauPlusMult
        /**
         * @param tauPlus Decay rate for r1 trace.
         */
        set(tauPlus) {
            this.tauPlusMult = 1 / tauPlus
        }

    var tauXDecay: Double
        /**
         * @return Decay rate for r2 trace.
         */
        get() = 1 / tauXMult
        /**
         * @param tauX Decay rate for r2 trace.
         */
        set(tauX) {
            this.tauXMult = 1 / tauX
        }

    var tauNegDecay: Double
        /**
         * @return Decay rate for o1 trace.
         */
        get() = 1 / tauNegMult
        /**
         * @param tauNeg Decay rate for o1 trace.
         */
        set(tauNeg) {
            this.tauNegMult = 1 / tauNeg
        }

    var tauYDecay: Double
        /**
         * @return Decay rate for o2 trace.
         */
        get() = 1 / tauYMult
        /**
         * @param tauY Decay rate for o2 trace.
         */
        set(tauY) {
            this.tauYMult = 1 / tauY
        }
}
